File: backend/src/main/java/com/decmoe47/todo/model/vo/R.java
package com.decmoe47.todo.model.vo;

import com.decmoe47.todo.constant.enums.ErrorCodeEnum;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;

@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Getter
public class R<T> implements Serializable {

    private int code;
    private String message;

    @Nullable
    private T data;

    public static <T> R<T> ok() {
        return new R<>(0, "", null);
    }

    public static <T> R<T> ok(T data) {
        return new R<>(0, "", data);
    }

    public static <T> R<T> error(ErrorCodeEnum errCode) {
        return new R<>(errCode.getCode(), errCode.getMessage(), null);
    }

    public static <T> R<T> error(ErrorCodeEnum errCode, T data) {
        return new R<>(errCode.getCode(), errCode.getMessage(), data);
    }
}

File: backend/src/main/kotlin/com/decmoe47/todo/model/response/R.kt
package com.decmoe47.todo.model.response

import com.decmoe47.todo.constant.enums.ErrorCode
import kotlinx.serialization.Serializable

@Serializable
data class R<out T>(
    val code: Int = 0,
    val message: String? = "",
    val data: T? = null
) : java.io.Serializable {
    companion object {
        fun ok(): R<Unit> = R()

        fun <T> ok(data: T): R<T> = R(data = data)

        fun error(errCode: ErrorCode): R<Unit> = R(errCode.code, errCode.message)

        fun <T> error(errCode: ErrorCode, data: T): R<T> = R(errCode.code, errCode.message, data)
    }
}